### ISSUE TITLE
Argument list too long

### ISSUE THREAD

**Issue by woollybah** [bmx-ng/bmk#21]
_From @Leushenko on January 20, 2016 21:53_

Following error when compiling:

    sh: /usr/bin/gcc: Argument list too long
    Build Error: failed to compile (32256) /Users/alex/Projects/...<etc>

No further details offered by the compiler. This is at around 63% completion.

_Copied from original issue: bmx-ng/bcc#156_

**Comment by woollybah** [comment]
Interesting. I had the same issue whilst testing on AROS, but the Amiga's max command-line length is *much* shorter than Linux...
Looks like we can use @file, and dump the compiler options into a temp file during the build..

**Comment by woollybah** [comment]
_From @GWRon on January 21, 2016 10:3_

Please take into consideration that this "temp file" must use a individual file name, just to avoid problems in a multithreaded compilation process when writing to "args.tmp".

I do not know how fast an md5 hash is (compared to the rest) but maybe hashing the complete content of that tempfile will result in a unique file name. But even then thread 1 might want to remove a file which thread 2 is using. This is of course only possible if there is the possibility of exactly the same arguments (compiling the same file?).

A (super-stricted) Md5-Hash-Code could be found in my framework:
https://github.com/GWRon/Dig/blob/master/base.util.hashes.bmx

Just copy/paste it if you want to use such a thing.

--
Maybe it is enough to use a tmp file named after the sourcecode filename we compile ... would make it way easier :-) (but allows for clashes when not somehow using the absolute path to that file)

**Comment by woollybah** [comment]
md5 is too slow. I'll use a system API for this probably.

**Comment by woollybah** [comment]
_From @GWRon on January 21, 2016 10:41_

Just as I was interested in measuring it:
```BlitzMax
SuperStrict
Framework Brl.StandardIO
Import "base.util.hashes.bmx"

Global hash:String
Global hashString:string = ""
For local i:int = 0 to 300
	hashString :+ "Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua." 
Next


Global startTime:int = Millisecs()
For local i:int = 0 to 100
	hash = Hashes.MD5(hashString)
Next
Global endTime:int = Millisecs()

print "length of string: "+hashString.length
print "hash: " + hash
print "time: " + ((endTime - startTime)/100.0) +" ms"
```
Result:
```
Executing:md5time
length of string: 46655
hash: 31ec1dc4be15cdeafd394365dc1007b4
time: 0.569999993 ms
```
I assume the actual compilation takes longer than this hashing process of a 45kb sized string.

**Comment by woollybah** [comment]
Yeah. I still prefer an API solution. Just because we are programmers, doesn't necessitate us designing new wheels every time ;-) :black_small_square:

**Comment by woollybah** [comment]
Out of interest, what result do you get from:
```
getconf ARG_MAX
```

Thanks!

**Comment by woollybah** [comment]
_From @GWRon on January 21, 2016 15:17_

```
$ getconf ARG_MAX
2097152
```

Sorry for the delay, just came home to my linux box

**Comment by woollybah** [comment]
_From @GWRon on January 21, 2016 15:47_

For __Windows__ the limits are discussed there:
https://blogs.msdn.microsoft.com/oldnewthing/20031210-00/?p=41553

CreateProcess: 32767 characters
CMD.exe: 8192 characters
ShellExecute/Ex: 2048 characters
Environment: 32767 characters
_spawnv(): 1024 bytes

**Comment by Leushenko** [comment]
I feel like this is somehow related to [this](https://github.com/bmx-ng/bcc/issues/159).

As said over there, I've got a lot of arguably-redundant `Import` declarations in my project (because of wanting to explicitly import everything relevant rather than trusting hidden details). When I removed a few `Import "runtime/Util.bmx` from files at the toplevel that should be able to see it indirectly anyway, the "argument list too long" error went away.

Noticed this because `util` itself was appearing multiple times in the compile list (which could mean nothing, I don't know what that actually represents). Maybe there's something in the way it detects duplicate paths?
